pallets/proxy-keystore/src/lib.rs

```diff
@@ -0,0 +1,466 @@
+
+            let block_number = <frame_system::Pallet<T>>::block_number();
+
+            for key_hash in key_hashes {
+                Self::revoke_key(identity.clone(), key_hash.clone(), block_number.clone())?;
+
+                Self::deposit_event(
+                    Event::KeyRevoked(
+                        identity.clone(),
+                        key_hash.clone(),
+                        block_number.clone(),
+                    ),
+                )
+            }
+
+            Ok(())
+        }
+
+        #[pallet::weight(T::WeightInfo::set_deposit())]
+        pub fn set_deposit(origin: OriginFor<T>, new_deposit: T::Balance) -> DispatchResult {
+            // Ensure that the origin is council or root.
+            Self::ensure_admin_origin(origin)?;
+
+            // Set the new deposit.
+            Self::set_new_deposit(new_deposit)?;
+
+            // Deposit the event.
+            Self::deposit_event(
+                Event::DepositSet(new_deposit),
+            );
+
+            Ok(())
+        }
+    }
+
+    impl<T: Config> Pallet<T> {
+        /// Ensure that the origin of a call is either an Admin, which is configured in the runtime,
+        /// or Root.
+        // TODO(cdamian): How do we ensure that this came from democracy and not council?
+        fn ensure_admin_origin(origin: OriginFor<T>) -> DispatchResult {
+            T::AdminOrigin::try_origin(origin.clone())
+                .map(|_| ())
+                .or_else(ensure_root)?;
+            Ok(())
+        }
+
+        /// Reserve the key deposit for the provided account.
+        fn reserve_key_deposit(account_id: T::AccountId, key_deposit: T::Balance) -> DispatchResult {
+            T::Currency::reserve(&account_id, key_deposit)
+        }
+
+        /// Return the key deposit for the provided account.
+        fn return_key_deposit(account_id: T::AccountId, key_deposit: T::Balance) {
+            let _ = T::Currency::unreserve(&account_id, key_deposit);
+        }
+
+        /// Set the new deposit that will be used when adding keys.
+        fn set_new_deposit(new_deposit: T::Balance) -> DispatchResult {
+            <KeyDeposit<T>>::set(new_deposit);
+            Ok(())
+        }
+
+        /// Add a key to the `Keys` and `LastKeyByPurpose` storages if the following checks pass:
+        ///
+        /// - The account has enough funds to cover the `key_deposit`;
+        /// - A key with the same hash does not exist in the storages;
+        /// - An un-revoked key with the same purpose does not exist in the storages;
+        ///
+        /// The `key_deposit` is reserved upon success.
+        fn add_key(
+            account_id: T::AccountId,
+            add_key: AddKey<T::Hash>,
+            key_deposit: T::Balance,
+        ) -> DispatchResult {
+            Self::reserve_key_deposit(account_id.clone(), key_deposit)?;
+
+            let keys_iter = <Keys<T>>::iter_prefix(account_id.clone());
```

Comment:
Yes but in that case we won't be able to know if a non-revoked key with a different hash and the same purpose already exists.

Reply:
I do not think we need to care about that, users can have multiple different keys with the same purpose, associated to different nodes. For example the signing keys. For the signing keys we actually will not look at the `LastKeyByPurpose`, since the public key is passed along with the p2p message. We just check that the key is associated to the identity and hasn't been revoked. So it is fine to always override the LastKeyByPurpose when a new key is added.

Reply:
Oh, there was definitely a misunderstanding on my side. Thank you for clarifying!